### PR TITLE
feat: add server crypto

### PR DESCRIPTION
**Pull request by dguyen**
## Description

Currently we are required to ensure PII data is not passed around in search parameters and in the open for GDPR reasons.

Allowing us to encrypt and decrypt values with expiry dates will allow us to ensure this doesn't happen.

## Changes Made

- Added TPRC router for encryption and decryption methods

## Testing Performed

- Tested encrypting and decrypting data with and without `expiredAt`
- Tested via directly accessing API and also via trpc in react components
- Tested parsing en email search param in a page and decrypting it successfully

## Checklist

- [X] I have tested these changes locally and they work as expected.
- [X] I have followed the project's coding style guidelines.

### CODE REVIEW

File: packages/lib/constants/crypto.ts [hunk 1]
```diff
@@ -1 +1,3 @@
 export const DOCUMENSO_ENCRYPTION_KEY = process.env.NEXT_PRIVATE_ENCRYPTION_KEY;
+
+export const DOCUMENSO_ENCRYPTION_SECONDARY_KEY = process.env.NEXT_PRIVATE_ENCRYPTION_SECONDARY_KEY;
```
[the remark this answers is not in the document]
Comment (dguyen) [reply packages/lib/constants/crypto.ts:3]: What do you mean?

The default key is empty which will throw an error unless the user adds it in

[the remark this answers is not in the document]
Comment (Mythie) [reply packages/lib/constants/crypto.ts:3]: I mean we should also do an `assert(DOCUMENSO_ENCRYPTION_KEY !== DOCUMENSO_ENCRYPTION_SECONDARY_KEY)`

Just incase a self-hoster does something inherently dangerous

File: packages/trpc/server/crypto/router.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  }),
+
+  decrypt: procedure.input(ZDecryptDataMutationSchema).mutation(({ input }) => {
+    try {
+      return decryptData(input.data);
+    } catch {
+      // Never leak errors for crypto.
+      return null;
+    }
+  }),
```
[the remark this answers is not in the document]
Comment (dguyen) [reply packages/trpc/server/crypto/router.ts:24]: Yeah that was a concern I had

But I noticed the email verification PR is accessing it in client side

I guess we can remove and make it decrypted on the server side before passing it to the client

[the remark this answers is not in the document]
Comment (dguyen) [reply packages/trpc/server/crypto/router.ts:24]: But if you think about it, they can still access this indirectly so I don't think it matters?

[the remark this answers is not in the document]
Comment (Mythie) [reply packages/trpc/server/crypto/router.ts:24]: I think the indirection helps if we do it well enough 😛

[the remark this answers is not in the document]
Comment (Mythie) [reply packages/trpc/server/crypto/router.ts:24]: The other PR can have it returned in the error or as a redirect url so it still occurs on the server side and mitigates the need for client side shenanigans

File: packages/lib/constants/crypto.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dguyen) [reply packages/lib/constants/crypto.ts:3]: Updated to make sure:
- Keys can't be the same
- Keys can't be empty

Added a warning for when the primary key is set to the old default:
- "CAFEBABE"

**Review by Mythie — APPROVED** [review]

File: packages/lib/constants/crypto.ts [hunk 3]
```diff
@@ -1 +1,23 @@
...
+
+if (!DOCUMENSO_ENCRYPTION_KEY || !DOCUMENSO_ENCRYPTION_SECONDARY_KEY) {
+  throw new Error('Missing DOCUMENSO_ENCRYPTION_KEY or DOCUMENSO_ENCRYPTION_SECONDARY_KEY keys');
+}
+
+if (DOCUMENSO_ENCRYPTION_KEY === DOCUMENSO_ENCRYPTION_SECONDARY_KEY) {
+  throw new Error(
+    'DOCUMENSO_ENCRYPTION_KEY and DOCUMENSO_ENCRYPTION_SECONDARY_KEY cannot be equal',
```
Comment (catalinpit) [inline packages/lib/constants/crypto.ts:11]: Wouldn't this sound better as:

* `DOCUMENSO_ENCRYPTION_KEY and DOCUMENSO_ENCRYPTION_SECONDARY_KEY cannot be the same`

or

* `You shouldn't use the same DOCUMENSO_ENCRYPTION_KEY and DOCUMENSO_ENCRYPTION_SECONDARY_KEY. Please choose different values.`

File: packages/lib/constants/crypto.ts [hunk 4]
```diff
@@ -1 +1,23 @@
...
+  );
+}
+
+if (DOCUMENSO_ENCRYPTION_KEY === 'CAFEBABE') {
```
Comment (catalinpit) [inline packages/lib/constants/crypto.ts:15]: Wasn't the default value removed from the .env file in the lines above?

File: packages/lib/server-only/crypto/decrypt.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+  if (!result.success) {
+    return null;
+  }
+
+  if (result.data.expiresAt !== undefined && result.data.expiresAt < Date.now()) {
+    return null;
+  }
+
+  return result.data.data;
```
Comment (catalinpit) [inline packages/lib/server-only/crypto/decrypt.ts:32]: Probably this is not a mistake, but just to make sure, is it `result.data.data`? Is that the `data` field from `ZEncryptedDataSchema`?

**Review by catalinpit — COMMENTED** [review]
Nice one!

File: packages/lib/constants/crypto.ts (hunk 4 above)
Comment (dguyen) [reply packages/lib/constants/crypto.ts:15]: It is, but this would be for legacy instances

**Pull request merged**